### DISCUSSION TITLE
Why is resume called when now <= wait_until in EasyTrafficLight::immediately_stop_until

### DISCUSSION THREAD

**Discussion by rcywongaa** [osrf/rmf_core#324] [Q&A] [answered]
Why is `resume_cb` called when `now <= wait_until->value()` in [`immediately_stop_until`](https://github.com/open-rmf/rmf_ros2/blob/2fe08e328f543fe6a4e0853a60607b5b52015f2a/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/EasyTrafficLight.cpp#L180)?

According to the [method description](https://github.com/open-rmf/rmf_ros2/blob/main/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/TrafficLight.hpp#L222)
>     /// Immediately stop until the specified time.
>     /// \param[in] time
>     ///   The time to wait until
`resume` should only be called after the specified `time` has passed?

Similarly, [a few lines up](https://github.com/open-rmf/rmf_ros2/blob/2fe08e328f543fe6a4e0853a60607b5b52015f2a/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/EasyTrafficLight.cpp#L149), the call is ignored if `time >= now` which also seems to contradict the description.
>     /// If the specified time has already passed by the time this function is
>     /// called, then there is no need to stop.

Something similar happens also in [`TrafficLight.cpp`](https://github.com/open-rmf/rmf_ros2/blob/main/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/TrafficLight.cpp#L931) which I also don't quite understand...

**Comment by mxgrey** [comment, marked as answer]
> Something similar happens also in [`TrafficLight.cpp`](https://github.com/open-rmf/rmf_ros2/blob/fe36afe5244d4aa267f879f63f2f48a8e4b3bafa/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/TrafficLight.cpp#L931) which I also don't quite understand...

This resume is correct, because it gets triggered when the traffic light had told the robot to wait until a certain time, but now the traffic light's plan has changed so the robot should continue. I'll add a comment into this code to make that more clear.

> Why is resume_cb called when now <= wait_until->value() in [`immediately_stop_until`](https://github.com/open-rmf/rmf_ros2/blob/2fe08e328f543fe6a4e0853a60607b5b52015f2a/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/EasyTrafficLight.cpp#L180)?

I agree this looks wrong. I'm not sure why this has gone unnoticed, although I suppose this situation just doesn't come up very often because the immediate stops are much less common than stopping at a checkpoint. I'll test out what happens when we flip this inequality.

On a separate note, the recommended place for RMF-related discussions is now in the `open-rmf` org at https://github.com/open-rmf/rmf/discussions

**Reply by rcywongaa** [reply]
Apologies, not sure how I missed the other discussion page... I'll be sure to post it there next time.
Thanks for the clarification as well!

**Reply by rcywongaa** [reply]
I did a bit of prodding around..  It seems the `immediately_stop_until` value taken from `wp.time()` in [this line](https://github.com/open-rmf/rmf_ros2/blob/2fe08e328f543fe6a4e0853a60607b5b52015f2a/rmf_fleet_adapter/src/rmf_fleet_adapter/agv/TrafficLight.cpp#L787) always gave me a time in the past, i.e. `plan_` that was passed to this `update_timing()` contained waypoints with times in the past.  This probably deserves a question on it's own... but how is the waypoint time from the plan generated?  Is the velocity of the AGV inferred somehow?

**Reply by mxgrey** [reply]
> this line always gave me a time in the past

That's normal, and to be expected pretty often. That effectively means that the robot does not need to come to an immediate stop.

To elaborate on what's going on there, `if (!wp.graph_index().has_value())` means the plan was generated from a start point that was in-between checkpoints. If the `time` value of a mid-lane plan waypoint is later than the current time, then that means the planner wants the robot to come to an immediate stop. If the mid-lane plan waypoint is earlier than the current time, then no immediate stop is needed.

We could probably short-circuit some of the logic by checking whether the `time` value is later than `now` before issuing the command to immediately stop, but the assumption is that the implementation of the Traffic Light command handle will filter this out anyway.

> how is the waypoint time from the plan generated? Is the velocity of the AGV inferred somehow?

When you instantiate a traffic light, you have to [pass in the vehicle traits of your robot](https://github.com/open-rmf/rmf_ros2/blob/fe36afe5244d4aa267f879f63f2f48a8e4b3bafa/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/Adapter.hpp#L174) which includes nominal velocity and nominal acceleration properties.
